File: backend/app/schemas/bug_report.py
from datetime import datetime
from uuid import UUID
from typing import Optional

from pydantic import Field

from app.schemas.base import Schema


class CaptchaChallenge(Schema):
    """Math CAPTCHA challenge for anonymous reporters."""

    token: str
    question: str


class BugReportCreate(Schema):
    """Payload for submitting a bug report."""

    message: str = Field(..., min_length=10, max_length=4000)
    page_url: Optional[str] = Field(None, max_length=2048)
    user_agent: Optional[str] = Field(None, max_length=512)
    # Required only for anonymous submissions
    captcha_token: Optional[str] = None
    captcha_answer: Optional[str] = None


class BugReportRead(Schema):
    """Admin-visible bug report."""

    id: UUID
    user_id: Optional[UUID] = None
    message: str
    page_url: Optional[str] = None
    user_agent: Optional[str] = None
    created_at: datetime

File: frontend/src/api/bugReport.ts
import { apiFetch } from "./client";

export interface CaptchaChallenge {
  token: string;
  question: string;
}

export interface BugReportPayload {
  message: string;
  page_url?: string;
  user_agent?: string;
  captcha_token?: string;
  captcha_answer?: string;
}

export interface BugReportRead {
  id: string;
  user_id: string | null;
  message: string;
  page_url: string | null;
  user_agent: string | null;
  created_at: string;
}

export function getCaptcha(): Promise<CaptchaChallenge> {
  return apiFetch<CaptchaChallenge>("/bug-reports/captcha", { method: "GET" });
}

export function submitBugReport(payload: BugReportPayload): Promise<BugReportRead> {
  return apiFetch<BugReportRead>("/bug-reports", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}
